plugins/perspective_api/newhelm/annotators/perspective_api.py

```diff
@@ -127,33 +137,73 @@ def _process_analyze_comment_response(self, response: Dict) -> Dict[str, float]:
         return flattened
 
 
-def _batch_execute_requests(client: discovery.Resource, requests: List) -> List:
+def _batch_execute_requests(
+    client: discovery.Resource, requests: List, num_retries: int, rng: random.Random
+) -> List:
     """Wrapper around Google's batch API.
 
     This can give significant speedup. For example for PerspectiveAPI, batching
     25 requests is about 15x faster than doing each as separate calls.
     https://googleapis.github.io/google-api-python-client/docs/batch.html
     """
 
-    batch_request: BatchHttpRequest = client.new_batch_http_request()
+    errors = [None] * len(requests)
     responses: List[Dict] = [{}] * len(requests)
 
     def _callback(request_id: str, response: Dict, error: HttpError):
+        index = int(request_id)
         if error:
-            raise error
-        responses[int(request_id)] = response
-
-    for i, request in enumerate(requests):
-        batch_request.add(
-            request=request,
-            request_id=str(i),
-            callback=_callback,
-        )
-    # Fills up responses using the callback
-    batch_request.execute()
+            errors[index] = error
+        else:
+            # Clear any past errors
+            errors[index] = None
+        responses[index] = response
+
+    # Keep track of what requests have not yet successfully gotten a response
+    needs_call = list(range(len(requests)))
+    for retry_count in range(num_retries + 1):
+        if retry_count > 0:
+            # Perform exponential backoff
+            sleep_amount = rng.random() * 2**retry_count
+            logging.info("Performing exponential backoff. Sleeping:", sleep_amount)
+            time.sleep(sleep_amount)
+
+        # Build up a batch
+        batch_request: BatchHttpRequest = client.new_batch_http_request()
+        for i in needs_call:
+            batch_request.add(
+                request=requests[i],
+                request_id=str(i),
+                callback=_callback,
+            )
+        # Fills up responses using the callback
+        batch_request.execute()
+
+        # Figure out which requests need to be tried again.
+        next_round_needs_call: List[int] = []
+        fatal_errors: List[HttpError] = []
+        for i in needs_call:
+            error = errors[i]
+            if error is not None:
+                if _is_retriable(error):
+                    next_round_needs_call.append(i)
+                else:
+                    fatal_errors.append(error)
+        if fatal_errors:
+            # Just use the first one as an example.
+            raise fatal_errors[0]
+        if not next_round_needs_call:
+            break
+        needs_call = next_round_needs_call
     return responses
 
 
+def _is_retriable(error: HttpError) -> bool:
```

Comment:
FWIW, any 5xx reply can be retried, but I'm fine starting here.

Reply:
Updated.